{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ad941b8a-ca52-498d-a189-8b80a36d34d1",
   "metadata": {},
   "source": [
    "## Sbatch SLURM scripting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "957cdf4b-23ba-4568-a949-eb982a0965a6",
   "metadata": {},
   "outputs": [],
   "source": [
    "cd scripts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "9fa0690e-865f-44e1-8dc1-0fa7cc2e326c",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "On branch ziwei\n",
      "Changes not staged for commit:\n",
      "  (use \"git add <file>...\" to update what will be committed)\n",
      "  (use \"git restore <file>...\" to discard changes in working directory)\n",
      "\tmodified:   ../analyses/zliu_run_accessions.txt\n",
      "\tmodified:   ziwei_read_QC.sh\n",
      "\n",
      "Untracked files:\n",
      "  (use \"git add <file>...\" to include in what will be committed)\n",
      "\t../analyses/Day1afternoon.ipynb\n",
      "\t../analyses/R_studio_abundances_heatmaps.Rmd\n",
      "\t../analyses/R_studio_abundances_heatmaps.nb.html\n",
      "\t../analyses/Untitled.ipynb\n",
      "\t../analyses/blastn/\n",
      "\t../analyses/blastn_para/\n",
      "\t../analyses/bowtie/\n",
      "\t../analyses/history.txt\n",
      "\t../analyses/kraken2/\n",
      "\t../analyses/krona/\n",
      "\t../analyses/test/\n",
      "\t../analyses/zliu-sample-sub-set_multiqc_report.html\n",
      "\t../analyses/zliu-sample-sub-set_multiqc_report_data/\n",
      "\t../analyses/zliu_flash2.log\n",
      "\t../blastn_parallel.sh\n",
      "\t../data/blast_db.ndb\n",
      "\t../data/blast_db.nhr\n",
      "\t../data/blast_db.nin\n",
      "\t../data/blast_db.not\n",
      "\t../data/blast_db.nsq\n",
      "\t../data/blast_db.ntf\n",
      "\t../data/blast_db.nto\n",
      "\t../data/blast_db/\n",
      "\t../data/bowtie2_DBs/\n",
      "\t../data/merged_pairs/\n",
      "\t../data/para_computing/\n",
      "\t../data/reference_seqs/\n",
      "\t../data/slurm.23319766.1.er\n",
      "\t../data/sra_fastq/\n",
      "\t../data/subset/\n",
      "\t.kreport2krona_sbatch_array.sh.swp\n",
      "\tblastn_parallel.sh\n",
      "\tblastn_sbatch_array.sh\n",
      "\tkraken2_bracken_sbatch_array.sh\n",
      "\tkraken2_sbatch_array.sh\n",
      "\tkreport2krona_sbatch_array.sh\n",
      "\tkreport2krona_sbatch_array_1.sh\n",
      "\toutputs/\n",
      "\t../snakemake/\n",
      "\n",
      "no changes added to commit (use \"git add\" and/or \"git commit -a\")\n"
     ]
    }
   ],
   "source": [
    "git status"
   ]
  },
  {
   "cell_type": "raw",
   "id": "8f38238c-dbc9-430b-a0d1-9dd0c93ed5ab",
   "metadata": {},
   "source": [
    "git add generic_sbatch_array.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "d6a5cea2-1050-4f1b-afbe-6e97e2cdda56",
   "metadata": {
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "source": [
    "git commit -m'sbatch_array'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "55d573a6-e581-47ec-9ad8-869eb2ad0a96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash_create_sample_collab_sqlite_db.ipynb  kraken2_sbatch_array.sh\n",
      "blastn_parallel.sh\t\t\t   kreport2krona_sbatch_array_1.sh\n",
      "blastn_sbatch_array.sh\t\t\t   kreport2krona_sbatch_array.sh\n",
      "generic_sbatch_array.sh\t\t\t   outputs\n",
      "kraken2_bracken_sbatch_array.sh\t\t   ziwei_read_QC.sh\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e8d9441-4edd-48f5-ae32-65a7c80038ee",
   "metadata": {},
   "source": [
    "## sbatch application: BLASTn of mini reads subsets verse refseq viral database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "83848ee4-d37a-49b8-b710-9ada3d6763f3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blastn\t\t\t\t      R_studio_abundances_heatmaps.Rmd\n",
      "blastn_para\t\t\t      sample_collab.db\n",
      "bowtie\t\t\t\t      sample_collab_db_schema.sql\n",
      "data\t\t\t\t      test\n",
      "Day1afternoon.ipynb\t\t      Untitled.ipynb\n",
      "history.txt\t\t\t      zliu_flash2.log\n",
      "kraken2\t\t\t\t      zliu_run_accessions.txt\n",
      "krona\t\t\t\t      zliu-sample-sub-set_multiqc_report_data\n",
      "outputs\t\t\t\t      zliu-sample-sub-set_multiqc_report.html\n",
      "R_studio_abundances_heatmaps.nb.html\n"
     ]
    }
   ],
   "source": [
    "cd ../analyses\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ac07a238-3472-40ba-aeb4-322b14d429c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "module load seqkit blast"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "f2c24861-3765-4dcb-b4d5-33083b725f6a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USAGE\n",
      "  makeblastdb [-h] [-help] [-in input_file] [-input_type type]\n",
      "    -dbtype molecule_type [-title database_title] [-parse_seqids]\n",
      "    [-hash_index] [-mask_data mask_data_files] [-mask_id mask_algo_ids]\n",
      "    [-mask_desc mask_algo_descriptions] [-gi_mask]\n",
      "    [-gi_mask_name gi_based_mask_names] [-out database_name]\n",
      "    [-blastdb_version version] [-max_file_sz number_of_bytes]\n",
      "    [-logfile File_Name] [-taxid TaxID] [-taxid_map TaxIDMapFile] [-version]\n",
      "\n",
      "DESCRIPTION\n",
      "   Application to create BLAST databases, version 2.12.0+\n",
      "\n",
      "REQUIRED ARGUMENTS\n",
      " -dbtype <String, `nucl', `prot'>\n",
      "   Molecule type of target db\n",
      "\n",
      "OPTIONAL ARGUMENTS\n",
      " -h\n",
      "   Print USAGE and DESCRIPTION;  ignore all other parameters\n",
      " -help\n",
      "   Print USAGE, DESCRIPTION and ARGUMENTS; ignore all other parameters\n",
      " -version\n",
      "   Print version number;  ignore other arguments\n",
      "\n",
      " *** Input options\n",
      " -in <File_In>\n",
      "   Input file/database name\n",
      "   Default = `-'\n",
      " -input_type <String, `asn1_bin', `asn1_txt', `blastdb', `fasta'>\n",
      "   Type of the data specified in input_file\n",
      "   Default = `fasta'\n",
      "\n",
      " *** Configuration options\n",
      " -title <String>\n",
      "   Title for BLAST database\n",
      "   Default = input file name provided to -in argument\n",
      " -parse_seqids\n",
      "   Option to parse seqid for FASTA input if set, for all other input types\n",
      "   seqids are parsed automatically\n",
      " -hash_index\n",
      "   Create index of sequence hash values.\n",
      "\n",
      " *** Sequence masking options\n",
      " -mask_data <String>\n",
      "   Comma-separated list of input files containing masking data as produced by\n",
      "   NCBI masking applications (e.g. dustmasker, segmasker, windowmasker)\n",
      " -mask_id <String>\n",
      "   Comma-separated list of strings to uniquely identify the masking algorithm\n",
      "    * Requires:  mask_data\n",
      "    * Incompatible with:  gi_mask\n",
      " -mask_desc <String>\n",
      "   Comma-separated list of free form strings to describe the masking algorithm\n",
      "   details\n",
      "    * Requires:  mask_id\n",
      " -gi_mask\n",
      "   Create GI indexed masking data.\n",
      "    * Requires:  parse_seqids\n",
      "    * Incompatible with:  mask_id\n",
      " -gi_mask_name <String>\n",
      "   Comma-separated list of masking data output files.\n",
      "    * Requires:  mask_data, gi_mask\n",
      "\n",
      " *** Output options\n",
      " -out <String>\n",
      "   Name of BLAST database to be created\n",
      "   Default = input file name provided to -in argumentRequired if multiple\n",
      "   file(s)/database(s) are provided as input\n",
      " -blastdb_version <Integer, 4..5>\n",
      "   Version of BLAST database to be created\n",
      "   Default = `5'\n",
      " -max_file_sz <String>\n",
      "   Maximum file size for BLAST database files\n",
      "   Default = `1GB'\n",
      " -logfile <File_Out>\n",
      "   File to which the program log should be redirected\n",
      "\n",
      " *** Taxonomy options\n",
      " -taxid <Integer, >=0>\n",
      "   Taxonomy ID to assign to all sequences\n",
      "    * Incompatible with:  taxid_map\n",
      " -taxid_map <File_In>\n",
      "   Text file mapping sequence IDs to taxonomy IDs.\n",
      "   Format:<SequenceId> <TaxonomyId><newline>\n",
      "    * Requires:  parse_seqids\n",
      "    * Incompatible with:  taxid\n",
      "\n",
      "\n",
      "\n",
      "Building a new DB, current time: 06/19/2022 16:23:01\n",
      "New DB name:   /shared/ifbstor1/projects/form_2022_19/ziwei/sars2copath/data/blast_db\n",
      "New DB title:  viral_db\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /shared/ifbstor1/projects/form_2022_19/ziwei/sars2copath/data/blast_db\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 14813 sequences in 9.01113 seconds.\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "makeblastdb -help # create a blast database using the refseq viral genomes\n",
    "\n",
    "makeblastdb -in /shared/projects/form_2022_19/refseq/viral.genomic.fna -out ../data/blast_db -title viral_db  -dbtype 'nucl'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "178da5ee-f795-4a3e-b544-005e2d021aa3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "blast_db.ndb  blast_db.nin  blast_db.nsq  blast_db.nto\n",
      "blast_db.nhr  blast_db.not  blast_db.ntf\n"
     ]
    }
   ],
   "source": [
    "cd ../data/blast_db\n",
    "ls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "bb3b9edb-740c-4ef2-9c71-e7356f46d51a",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true,
     "source_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "USAGE\n",
      "  blastn [-h] [-help] [-import_search_strategy filename]\n",
      "    [-export_search_strategy filename] [-task task_name] [-db database_name]\n",
      "    [-dbsize num_letters] [-gilist filename] [-seqidlist filename]\n",
      "    [-negative_gilist filename] [-negative_seqidlist filename]\n",
      "    [-taxids taxids] [-negative_taxids taxids] [-taxidlist filename]\n",
      "    [-negative_taxidlist filename] [-entrez_query entrez_query]\n",
      "    [-db_soft_mask filtering_algorithm] [-db_hard_mask filtering_algorithm]\n",
      "    [-subject subject_input_file] [-subject_loc range] [-query input_file]\n",
      "    [-out output_file] [-evalue evalue] [-word_size int_value]\n",
      "    [-gapopen open_penalty] [-gapextend extend_penalty]\n",
      "    [-perc_identity float_value] [-qcov_hsp_perc float_value]\n",
      "    [-max_hsps int_value] [-xdrop_ungap float_value] [-xdrop_gap float_value]\n",
      "    [-xdrop_gap_final float_value] [-searchsp int_value]\n",
      "    [-sum_stats bool_value] [-penalty penalty] [-reward reward] [-no_greedy]\n",
      "    [-min_raw_gapped_score int_value] [-template_type type]\n",
      "    [-template_length int_value] [-dust DUST_options]\n",
      "    [-filtering_db filtering_database]\n",
      "    [-window_masker_taxid window_masker_taxid]\n",
      "    [-window_masker_db window_masker_db] [-soft_masking soft_masking]\n",
      "    [-ungapped] [-culling_limit int_value] [-best_hit_overhang float_value]\n",
      "    [-best_hit_score_edge float_value] [-subject_besthit]\n",
      "    [-window_size int_value] [-off_diagonal_range int_value]\n",
      "    [-use_index boolean] [-index_name string] [-lcase_masking]\n",
      "    [-query_loc range] [-strand strand] [-parse_deflines] [-outfmt format]\n",
      "    [-show_gis] [-num_descriptions int_value] [-num_alignments int_value]\n",
      "    [-line_length line_length] [-html] [-sorthits sort_hits]\n",
      "    [-sorthsps sort_hsps] [-max_target_seqs num_sequences]\n",
      "    [-num_threads int_value] [-mt_mode int_value] [-remote] [-version]\n",
      "\n",
      "DESCRIPTION\n",
      "   Nucleotide-Nucleotide BLAST 2.12.0+\n",
      "\n",
      "OPTIONAL ARGUMENTS\n",
      " -h\n",
      "   Print USAGE and DESCRIPTION;  ignore all other parameters\n",
      " -help\n",
      "   Print USAGE, DESCRIPTION and ARGUMENTS; ignore all other parameters\n",
      " -version\n",
      "   Print version number;  ignore other arguments\n",
      "\n",
      " *** Input query options\n",
      " -query <File_In>\n",
      "   Input file name\n",
      "   Default = `-'\n",
      " -query_loc <String>\n",
      "   Location on the query sequence in 1-based offsets (Format: start-stop)\n",
      " -strand <String, `both', `minus', `plus'>\n",
      "   Query strand(s) to search against database/subject\n",
      "   Default = `both'\n",
      "\n",
      " *** General search options\n",
      " -task <String, Permissible values: 'blastn' 'blastn-short' 'dc-megablast'\n",
      "                'megablast' 'rmblastn' >\n",
      "   Task to execute\n",
      "   Default = `megablast'\n",
      " -db <String>\n",
      "   BLAST database name\n",
      "    * Incompatible with:  subject, subject_loc\n",
      " -out <File_Out>\n",
      "   Output file name\n",
      "   Default = `-'\n",
      " -evalue <Real>\n",
      "   Expectation value (E) threshold for saving hits \n",
      "   Default = `10'\n",
      " -word_size <Integer, >=4>\n",
      "   Word size for wordfinder algorithm (length of best perfect match)\n",
      " -gapopen <Integer>\n",
      "   Cost to open a gap\n",
      " -gapextend <Integer>\n",
      "   Cost to extend a gap\n",
      " -penalty <Integer, <=0>\n",
      "   Penalty for a nucleotide mismatch\n",
      " -reward <Integer, >=0>\n",
      "   Reward for a nucleotide match\n",
      " -use_index <Boolean>\n",
      "   Use MegaBLAST database index\n",
      "   Default = `false'\n",
      " -index_name <String>\n",
      "   MegaBLAST database index name (deprecated; use only for old style indices)\n",
      "\n",
      " *** BLAST-2-Sequences options\n",
      " -subject <File_In>\n",
      "   Subject sequence(s) to search\n",
      "    * Incompatible with:  db, gilist, seqidlist, negative_gilist,\n",
      "   negative_seqidlist, taxids, taxidlist, negative_taxids, negative_taxidlist,\n",
      "   db_soft_mask, db_hard_mask\n",
      " -subject_loc <String>\n",
      "   Location on the subject sequence in 1-based offsets (Format: start-stop)\n",
      "    * Incompatible with:  db, gilist, seqidlist, negative_gilist,\n",
      "   negative_seqidlist, taxids, taxidlist, negative_taxids, negative_taxidlist,\n",
      "   db_soft_mask, db_hard_mask, remote\n",
      "\n",
      " *** Formatting options\n",
      " -outfmt <String>\n",
      "   alignment view options:\n",
      "     0 = Pairwise,\n",
      "     1 = Query-anchored showing identities,\n",
      "     2 = Query-anchored no identities,\n",
      "     3 = Flat query-anchored showing identities,\n",
      "     4 = Flat query-anchored no identities,\n",
      "     5 = BLAST XML,\n",
      "     6 = Tabular,\n",
      "     7 = Tabular with comment lines,\n",
      "     8 = Seqalign (Text ASN.1),\n",
      "     9 = Seqalign (Binary ASN.1),\n",
      "    10 = Comma-separated values,\n",
      "    11 = BLAST archive (ASN.1),\n",
      "    12 = Seqalign (JSON),\n",
      "    13 = Multiple-file BLAST JSON,\n",
      "    14 = Multiple-file BLAST XML2,\n",
      "    15 = Single-file BLAST JSON,\n",
      "    16 = Single-file BLAST XML2,\n",
      "    17 = Sequence Alignment/Map (SAM),\n",
      "    18 = Organism Report\n",
      "   \n",
      "   Options 6, 7, 10 and 17 can be additionally configured to produce\n",
      "   a custom format specified by space delimited format specifiers,\n",
      "   or in the case of options 6, 7, and 10, by a token specified\n",
      "   by the delim keyword. E.g.: \"17 delim=@ qacc sacc score\".\n",
      "   The delim keyword must appear after the numeric output format\n",
      "   specification.\n",
      "   The supported format specifiers for options 6, 7 and 10 are:\n",
      "   \t    qseqid means Query Seq-id\n",
      "   \t       qgi means Query GI\n",
      "   \t      qacc means Query accesion\n",
      "   \t   qaccver means Query accesion.version\n",
      "   \t      qlen means Query sequence length\n",
      "   \t    sseqid means Subject Seq-id\n",
      "   \t sallseqid means All subject Seq-id(s), separated by a ';'\n",
      "   \t       sgi means Subject GI\n",
      "   \t    sallgi means All subject GIs\n",
      "   \t      sacc means Subject accession\n",
      "   \t   saccver means Subject accession.version\n",
      "   \t   sallacc means All subject accessions\n",
      "   \t      slen means Subject sequence length\n",
      "   \t    qstart means Start of alignment in query\n",
      "   \t      qend means End of alignment in query\n",
      "   \t    sstart means Start of alignment in subject\n",
      "   \t      send means End of alignment in subject\n",
      "   \t      qseq means Aligned part of query sequence\n",
      "   \t      sseq means Aligned part of subject sequence\n",
      "   \t    evalue means Expect value\n",
      "   \t  bitscore means Bit score\n",
      "   \t     score means Raw score\n",
      "   \t    length means Alignment length\n",
      "   \t    pident means Percentage of identical matches\n",
      "   \t    nident means Number of identical matches\n",
      "   \t  mismatch means Number of mismatches\n",
      "   \t  positive means Number of positive-scoring matches\n",
      "   \t   gapopen means Number of gap openings\n",
      "   \t      gaps means Total number of gaps\n",
      "   \t      ppos means Percentage of positive-scoring matches\n",
      "   \t    frames means Query and subject frames separated by a '/'\n",
      "   \t    qframe means Query frame\n",
      "   \t    sframe means Subject frame\n",
      "   \t      btop means Blast traceback operations (BTOP)\n",
      "   \t    staxid means Subject Taxonomy ID\n",
      "   \t  ssciname means Subject Scientific Name\n",
      "   \t  scomname means Subject Common Name\n",
      "   \tsblastname means Subject Blast Name\n",
      "   \t sskingdom means Subject Super Kingdom\n",
      "   \t   staxids means unique Subject Taxonomy ID(s), separated by a ';'\n",
      "   \t\t\t (in numerical order)\n",
      "   \t sscinames means unique Subject Scientific Name(s), separated by a ';'\n",
      "   \t scomnames means unique Subject Common Name(s), separated by a ';'\n",
      "   \tsblastnames means unique Subject Blast Name(s), separated by a ';'\n",
      "   \t\t\t (in alphabetical order)\n",
      "   \tsskingdoms means unique Subject Super Kingdom(s), separated by a ';'\n",
      "   \t\t\t (in alphabetical order) \n",
      "   \t    stitle means Subject Title\n",
      "   \tsalltitles means All Subject Title(s), separated by a '<>'\n",
      "   \t   sstrand means Subject Strand\n",
      "   \t     qcovs means Query Coverage Per Subject\n",
      "   \t   qcovhsp means Query Coverage Per HSP\n",
      "   \t    qcovus means Query Coverage Per Unique Subject (blastn only)\n",
      "   When not provided, the default value is:\n",
      "   'qaccver saccver pident length mismatch gapopen qstart qend sstart send\n",
      "   evalue bitscore', which is equivalent to the keyword 'std'\n",
      "   The supported format specifier for option 17 is:\n",
      "   \t        SQ means Include Sequence Data\n",
      "   \t        SR means Subject as Reference Seq\n",
      "   Default = `0'\n",
      " -show_gis\n",
      "   Show NCBI GIs in deflines?\n",
      " -num_descriptions <Integer, >=0>\n",
      "   Number of database sequences to show one-line descriptions for\n",
      "   Not applicable for outfmt > 4\n",
      "   Default = `500'\n",
      "    * Incompatible with:  max_target_seqs\n",
      " -num_alignments <Integer, >=0>\n",
      "   Number of database sequences to show alignments for\n",
      "   Default = `250'\n",
      "    * Incompatible with:  max_target_seqs\n",
      " -line_length <Integer, >=1>\n",
      "   Line length for formatting alignments\n",
      "   Not applicable for outfmt > 4\n",
      "   Default = `60'\n",
      " -html\n",
      "   Produce HTML output?\n",
      " -sorthits <Integer, (>=0 and =<4)>\n",
      "   Sorting option for hits:\n",
      "   alignment view options:\n",
      "     0 = Sort by evalue,\n",
      "     1 = Sort by bit score,\n",
      "     2 = Sort by total score,\n",
      "     3 = Sort by percent identity,\n",
      "     4 = Sort by query coverage\n",
      "   Not applicable for outfmt > 4\n",
      " -sorthsps <Integer, (>=0 and =<4)>\n",
      "   Sorting option for hps:\n",
      "     0 = Sort by hsp evalue,\n",
      "     1 = Sort by hsp score,\n",
      "     2 = Sort by hsp query start,\n",
      "     3 = Sort by hsp percent identity,\n",
      "     4 = Sort by hsp subject start\n",
      "   Not applicable for outfmt != 0\n",
      "\n",
      " *** Query filtering options\n",
      " -dust <String>\n",
      "   Filter query sequence with DUST (Format: 'yes', 'level window linker', or\n",
      "   'no' to disable)\n",
      "   Default = `20 64 1'\n",
      " -filtering_db <String>\n",
      "   BLAST database containing filtering elements (i.e.: repeats)\n",
      " -window_masker_taxid <Integer>\n",
      "   Enable WindowMasker filtering using a Taxonomic ID\n",
      " -window_masker_db <String>\n",
      "   Enable WindowMasker filtering using this repeats database.\n",
      " -soft_masking <Boolean>\n",
      "   Apply filtering locations as soft masks\n",
      "   Default = `true'\n",
      " -lcase_masking\n",
      "   Use lower case filtering in query and subject sequence(s)?\n",
      "\n",
      " *** Restrict search or results\n",
      " -gilist <String>\n",
      "   Restrict search of database to list of GIs\n",
      "    * Incompatible with:  seqidlist, taxids, taxidlist, negative_gilist,\n",
      "   negative_seqidlist, negative_taxids, negative_taxidlist, remote, subject,\n",
      "   subject_loc\n",
      " -seqidlist <String>\n",
      "   Restrict search of database to list of SeqIDs\n",
      "    * Incompatible with:  gilist, taxids, taxidlist, negative_gilist,\n",
      "   negative_seqidlist, negative_taxids, negative_taxidlist, remote, subject,\n",
      "   subject_loc\n",
      " -negative_gilist <String>\n",
      "   Restrict search of database to everything except the specified GIs\n",
      "    * Incompatible with:  gilist, seqidlist, taxids, taxidlist,\n",
      "   negative_seqidlist, negative_taxids, negative_taxidlist, remote, subject,\n",
      "   subject_loc\n",
      " -negative_seqidlist <String>\n",
      "   Restrict search of database to everything except the specified SeqIDs\n",
      "    * Incompatible with:  gilist, seqidlist, taxids, taxidlist,\n",
      "   negative_gilist, negative_taxids, negative_taxidlist, remote, subject,\n",
      "   subject_loc\n",
      " -taxids <String>\n",
      "   Restrict search of database to include only the specified taxonomy IDs\n",
      "   (multiple IDs delimited by ',')\n",
      "    * Incompatible with:  gilist, seqidlist, taxidlist, negative_gilist,\n",
      "   negative_seqidlist, negative_taxids, negative_taxidlist, remote, subject,\n",
      "   subject_loc\n",
      " -negative_taxids <String>\n",
      "   Restrict search of database to everything except the specified taxonomy IDs\n",
      "   (multiple IDs delimited by ',')\n",
      "    * Incompatible with:  gilist, seqidlist, taxids, taxidlist,\n",
      "   negative_gilist, negative_seqidlist, negative_taxidlist, remote, subject,\n",
      "   subject_loc\n",
      " -taxidlist <String>\n",
      "   Restrict search of database to include only the specified taxonomy IDs\n",
      "    * Incompatible with:  gilist, seqidlist, taxids, negative_gilist,\n",
      "   negative_seqidlist, negative_taxids, negative_taxidlist, remote, subject,\n",
      "   subject_loc\n",
      " -negative_taxidlist <String>\n",
      "   Restrict search of database to everything except the specified taxonomy IDs\n",
      "    * Incompatible with:  gilist, seqidlist, taxids, taxidlist,\n",
      "   negative_gilist, negative_seqidlist, negative_taxids, remote, subject,\n",
      "   subject_loc\n",
      " -entrez_query <String>\n",
      "   Restrict search with the given Entrez query\n",
      "    * Requires:  remote\n",
      " -db_soft_mask <String>\n",
      "   Filtering algorithm ID to apply to the BLAST database as soft masking\n",
      "    * Incompatible with:  db_hard_mask, subject, subject_loc\n",
      " -db_hard_mask <String>\n",
      "   Filtering algorithm ID to apply to the BLAST database as hard masking\n",
      "    * Incompatible with:  db_soft_mask, subject, subject_loc\n",
      " -perc_identity <Real, 0..100>\n",
      "   Percent identity\n",
      " -qcov_hsp_perc <Real, 0..100>\n",
      "   Percent query coverage per hsp\n",
      " -max_hsps <Integer, >=1>\n",
      "   Set maximum number of HSPs per subject sequence to save for each query\n",
      " -culling_limit <Integer, >=0>\n",
      "   If the query range of a hit is enveloped by that of at least this many\n",
      "   higher-scoring hits, delete the hit\n",
      "    * Incompatible with:  best_hit_overhang, best_hit_score_edge\n",
      " -best_hit_overhang <Real, (>0 and <0.5)>\n",
      "   Best Hit algorithm overhang value (recommended value: 0.1)\n",
      "    * Incompatible with:  culling_limit\n",
      " -best_hit_score_edge <Real, (>0 and <0.5)>\n",
      "   Best Hit algorithm score edge value (recommended value: 0.1)\n",
      "    * Incompatible with:  culling_limit\n",
      " -subject_besthit\n",
      "   Turn on best hit per subject sequence\n",
      " -max_target_seqs <Integer, >=1>\n",
      "   Maximum number of aligned sequences to keep \n",
      "   (value of 5 or more is recommended)\n",
      "   Default = `500'\n",
      "    * Incompatible with:  num_descriptions, num_alignments\n",
      "\n",
      " *** Discontiguous MegaBLAST options\n",
      " -template_type <String, `coding', `coding_and_optimal', `optimal'>\n",
      "   Discontiguous MegaBLAST template type\n",
      "    * Requires:  template_length\n",
      " -template_length <Integer, Permissible values: '16' '18' '21' >\n",
      "   Discontiguous MegaBLAST template length\n",
      "    * Requires:  template_type\n",
      "\n",
      " *** Statistical options\n",
      " -dbsize <Int8>\n",
      "   Effective length of the database \n",
      " -searchsp <Int8, >=0>\n",
      "   Effective length of the search space\n",
      " -sum_stats <Boolean>\n",
      "   Use sum statistics\n",
      "\n",
      " *** Search strategy options\n",
      " -import_search_strategy <File_In>\n",
      "   Search strategy to use\n",
      "    * Incompatible with:  export_search_strategy\n",
      " -export_search_strategy <File_Out>\n",
      "   File name to record the search strategy used\n",
      "    * Incompatible with:  import_search_strategy\n",
      "\n",
      " *** Extension options\n",
      " -xdrop_ungap <Real>\n",
      "   X-dropoff value (in bits) for ungapped extensions\n",
      " -xdrop_gap <Real>\n",
      "   X-dropoff value (in bits) for preliminary gapped extensions\n",
      " -xdrop_gap_final <Real>\n",
      "   X-dropoff value (in bits) for final gapped alignment\n",
      " -no_greedy\n",
      "   Use non-greedy dynamic programming extension\n",
      " -min_raw_gapped_score <Integer>\n",
      "   Minimum raw gapped score to keep an alignment in the preliminary gapped and\n",
      "   traceback stages\n",
      " -ungapped\n",
      "   Perform ungapped alignment only?\n",
      " -window_size <Integer, >=0>\n",
      "   Multiple hits window size, use 0 to specify 1-hit algorithm\n",
      " -off_diagonal_range <Integer, >=0>\n",
      "   Number of off-diagonals to search for the 2nd hit, use 0 to turn off\n",
      "   Default = `0'\n",
      "\n",
      " *** Miscellaneous options\n",
      " -parse_deflines\n",
      "   Should the query and subject defline(s) be parsed?\n",
      " -num_threads <Integer, >=1>\n",
      "   Number of threads (CPUs) to use in the BLAST search\n",
      "   Default = `1'\n",
      "    * Incompatible with:  remote\n",
      " -mt_mode <Integer, (>=0 and =<1)>\n",
      "   Multi-thread mode to use in BLAST search:\n",
      "    0 (auto) split by database \n",
      "    1 split by queries\n",
      "   Default = `0'\n",
      "    * Requires:  num_threads\n",
      " -remote\n",
      "   Execute search remotely?\n",
      "    * Incompatible with:  gilist, seqidlist, taxids, taxidlist,\n",
      "   negative_gilist, negative_seqidlist, negative_taxids, negative_taxidlist,\n",
      "   subject_loc, num_threads\n",
      "\n"
     ]
    }
   ],
   "source": [
    "blastn -help"
   ]
  },
  {
   "cell_type": "raw",
   "id": "5375cbb0-b278-4ead-a96a-8a25c1f40b60",
   "metadata": {
    "tags": []
   },
   "source": [
    "seqkit head -n10 /shared/projects/form_2022_19/ziwei/sars2copath/data/merged_pairs/ERR6913297.flash.extendedFrags.fastq.gz > /shared/projects/form_2022_19/ziwei/sars2copath/data/subset/10_fastq.gz"
   ]
  },
  {
   "cell_type": "raw",
   "id": "b9da4856-c965-49e4-aa6e-faaff3db8934",
   "metadata": {},
   "source": [
    "seqkit head -n100 /shared/projects/form_2022_19/ziwei/sars2copath/data/merged_pairs/ERR6913297.flash.extendedFrags.fastq.gz > /shared/projects/form_2022_19/ziwei/sars2copath/data/subset/100_fastq.gz"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7a9b8638-771b-46ea-9958-49a883d2cb6f",
   "metadata": {},
   "source": [
    "seqkit head -n1000 /shared/projects/form_2022_19/ziwei/sars2copath/data/merged_pairs/ERR6913297.flash.extendedFrags.fastq.gz > /shared/projects/form_2022_19/ziwei/sars2copath/data/subset/1000_fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e13e6911-fd56-4d36-b3ab-021996993628",
   "metadata": {},
   "source": [
    "convert the fastq.gz to fasta format because the blastn only takes in fasta files"
   ]
  },
  {
   "cell_type": "raw",
   "id": "2eaab1e1-a4de-477f-b2fc-4fb119a107de",
   "metadata": {},
   "source": [
    "cd ../data/subset\n",
    "mkdir fasta"
   ]
  },
  {
   "cell_type": "raw",
   "id": "3733f8d1-6cec-4c80-8a9b-1ec284790bfa",
   "metadata": {},
   "source": [
    "seqkit fq2fa 10_fastq.gz -o fasta/10_fasta.gz"
   ]
  },
  {
   "cell_type": "raw",
   "id": "04529cc6-30cf-4ac2-8db0-d1603eb59e45",
   "metadata": {},
   "source": [
    "seqkit fq2fa 100_fastq.gz -o fasta/100_fasta.gz"
   ]
  },
  {
   "cell_type": "raw",
   "id": "85af33e5-94b8-4f3d-a093-5d0d1a9e30cc",
   "metadata": {},
   "source": [
    "seqkit fq2fa 1000_fastq.gz -o fasta/1000_fasta.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "954825f2-ce7b-4a21-92b5-cd4269c4f83c",
   "metadata": {},
   "source": [
    "write a unique sbatch script to blastn the 10, 100 & 1000 reads subset against the above created refseq viral database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "022c886a-b14d-4baa-9201-e7fa4fcedb0e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bash_create_sample_collab_sqlite_db.ipynb  kraken2_sbatch_array.sh\n",
      "blastn_parallel.sh\t\t\t   kreport2krona_sbatch_array_1.sh\n",
      "blastn_sbatch_array.sh\t\t\t   kreport2krona_sbatch_array.sh\n",
      "generic_sbatch_array.sh\t\t\t   outputs\n",
      "kraken2_bracken_sbatch_array.sh\t\t   ziwei_read_QC.sh\n"
     ]
    }
   ],
   "source": [
    "ls"
   ]
  },
  {
   "cell_type": "raw",
   "id": "13bbf9f4-2715-4ef4-964e-55408e888d57",
   "metadata": {},
   "source": [
    "[zliu@clust-slurm-client scripts]$ vim blastn_sbatch_array.sh\n",
    "\n",
    "#\n",
    "#SBATCH --partition=fast             # long, fast, etc.\n",
    "#SBATCH --ntasks=1                   # nb of *tasks* to be run in // (usually 1), this task can be multithreaded (see cpus-per-task)\n",
    "#SBATCH --nodes=1                    # nb of nodes to reserve for each task (usually 1)\n",
    "#SBATCH --cpus-per-task=1            # nb of cpu (in fact cores) to reserve for each task /!\\ job killed if commands below use more cores\n",
    "#SBATCH --mem=15GB                  # amount of RAM to reserve for the tasks /!\\ job killed if commands below use more RAM\n",
    "#SBATCH --time=0-0:15               # maximal wall clock duration (D-HH:MM) /!\\ job killed if commands below take more time than reservation\n",
    "#SBATCH -o ./outputs/slurm.%A.%a.out   # standard output (STDOUT) redirected to these files (with Job ID and array ID in file names)\n",
    "#SBATCH -e ./outputs/slurm.%A.%a.err   # standard error  (STDERR) redirected to these files (with Job ID and array ID in file names)\n",
    "# /!\\ Note that the ./outputs/ dir above needs to exist in the dir where script is submitted **prior** to submitting this script\n",
    "#SBATCH --array=1-8                # 1-N: clone this script in an array of N tasks: $SLURM_ARRAY_TASK_ID will take the value of 1,2,...,N\n",
    "#SBATCH --job-name=MedBioinfo        # name of the task as displayed in squeue & sacc, also encouraged as srun optional parameter\n",
    "#SBATCH --mail-type END              # when to send an email notiification (END = when the whole sbatch array is finished)\n",
    "#SBATCH --mail-user ziwei.liu@ki.se\n",
    "\n",
    "#################################################################\n",
    "# Preparing work (cd to work\n",
    "ing dir, get hold of input data, convert/un-compress input data when needed etc.)\n",
    "workdir=\"/shared/projects/form_2022_19/ziwei/sars2copath/analyses/blastn/\"\n",
    "datadir=\"/shared/projects/form_2022_19/ziwei/sars2copath/data/\"\n",
    "accnum_file=\"/shared/projects/form_2022_19/ziwei/file_of_acc_nums.txt\"\n",
    "\n",
    "echo START: `date`\n",
    "\n",
    "module load seqkit blast #as required\n",
    "\n",
    "mkdir -p ${workdir}      # -p because it creates all required dir levels **and** doesn't throw an error if the dir exists :)\n",
    "cd ${workdir}\n",
    "\n",
    "\n",
    "# because there are mutliple jobs running in // each output file needs to be made unique by post-fixing with $SLURM_ARRAY_TASK_ID and/or $accnum\n",
    "#output_file=\"${workdir}/1000_blastn_vs_viral_refseq.out\"\n",
    "\n",
    "#################################################################\n",
    "# Start work\n",
    "srun blastn -db ${datadir}/blast_db -num_threads ${SLURM_CPUS_PER_TASK} -query ${datadir}/subset/fasta/1000samples.fa -out ${workdir}/1000_blastn_vs_viral_refseq.blast -outfmt 6\n",
    "\n",
    "#################################################################\n",
    "# Clean up (eg delete temp files, compress output, recompress input etc)\n",
    "#srun gzip ${input_file}\n",
    "#srun gzip ${output_file}\n",
    "echo END: `date`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3b00d772-c068-4746-83a6-ed72b538c718",
   "metadata": {},
   "source": [
    "### sbtach parallel job arrays"
   ]
  },
  {
   "cell_type": "raw",
   "id": "3c179921-fc30-4144-a9c3-d2e963aa05fa",
   "metadata": {},
   "source": [
    "[zliu@clust-slurm-client scripts]$ vim blastn_parallel.sh"
   ]
  },
  {
   "cell_type": "raw",
   "id": "196cea60-1257-4e89-b163-d8592511865d",
   "metadata": {},
   "source": [
    "[zliu@clust-slurm-client scripts]$ vim blastn_parallel.sh\n",
    "\n",
    "#!/bin/bash\n",
    "#\n",
    "#SBATCH --partition=fast             # long, fast, etc.\n",
    "#SBATCH --ntasks=1                   # nb of *tasks* to be run in // (usually 1), this task can be multithreaded (see cpus-per-task)\n",
    "#SBATCH --nodes=1                    # nb of nodes to reserve for each task (usually 1)\n",
    "#SBATCH --cpus-per-task=4            # nb of cpu (in fact cores) to reserve for each task /!\\ job killed if commands below use more cores\n",
    "#SBATCH --mem=25GB                  # amount of RAM to reserve for the tasks /!\\ job killed if commands below use more RAM\n",
    "#SBATCH --time=0-01:00               # maximal wall clock duration (D-HH:MM) /!\\ job killed if commands below take more time than reservation\n",
    "#SBATCH -o ./outputs/slurm.%A.%a.out   # standard output (STDOUT) redirected to these files (with Job ID and array ID in file names)\n",
    "#SBATCH -e ./outputs/slurm.%A.%a.err   # standard error  (STDERR) redirected to these files (with Job ID and array ID in file names)\n",
    "# /!\\ Note that the ./outputs/ dir above needs to exist in the dir where script is submitted **prior** to submitting this script\n",
    "#SBATCH --array=1-8             # 1-N: clone this script in an array of N tasks: $SLURM_ARRAY_TASK_ID will take the value of 1,2,...,N\n",
    "#SBATCH --job-name=${accnum}        # name of the task as displayed in squeue & sacc, also encouraged as srun optional parameter\n",
    "#SBATCH --mail-type END              # when to send an email notiification (END = when the whole sbatch array is finished)\n",
    "#SBATCH --mail-user ziwei.liu@ki.com\n",
    "\n",
    "#################################################################\n",
    "# Preparing work (cd to working dir, get hold of input data, convert/un-compress input data when needed etc.)\n",
    "workdir=\"/shared/projects/form_2022_19/ziwei/sars2copath/analyses/blastn_para/\"\n",
    "datadir=\"/shared/projects/form_2022_19/ziwei/sars2copath/data/\"\n",
    "accnum_file=\"/shared/projects/form_2022_19/ziwei/sars2copath/analyses/zliu_run_accessions.txt\"\n",
    "\n",
    "echo START: `date`\n",
    "\n",
    "module load seqkit blast #as required\n",
    "\n",
    "mkdir -p ${workdir}      # -p because it creates all required dir levels **and** doesn't throw an error if the dir exists :)\n",
    "cd ${workdir}\n",
    "\n",
    "# this extracts the item number $SLURM_ARRAY_TASK_ID from the file of accnums\n",
    "accnum=$(sed -n \"$SLURM_ARRAY_TASK_ID\"p ${accnum_file})\n",
    "\n",
    "input_file=\"${datadir}merged_pairs/${accnum}.flash.extendedFrags.fastq.gz\"\n",
    "\n",
    "# alternatively, just extract the input file as the item number $SLURM_ARRAY_TASK_ID in the data dir listing\n",
    "# this alternative is less handy since we don't get hold of the isolated \"accnum\", which is very handy to name the srun step below :)\n",
    "#input_file=$(ls \"${datadir}/*.fastq.gz\" | sed -n ${SLURM_ARRAY_TASK_ID}p)\n",
    "\n",
    "# if the command below can't cope with compressed input\n",
    "# srun gunzip \"${input_file}.gz\"\n",
    "#convert fast\n",
    "srun seqkit fq2fa ${input_file} -o ${datadir}para_computing/${accnum}.fa\n",
    "\n",
    "# because there are mutliple jobs running in // each output file needs to be made unique by post-fixing with $SLURM_ARRAY_TASK_ID and/or $accnum\n",
    "output_file=\"${workdir}ABCjob.${accnum}.out\"\n",
    "\n",
    "#################################################################\n",
    "# Start work\n",
    "srun --job-name=${accnum} blastn -db ${datadir}/blast_db -num_threads ${SLURM_CPUS_PER_TASK} -query ${datadir}para_computing/${accnum}.fa -out ${output_file} -outfmt 6 -max_target_seqs 5\n",
    "\n",
    "#################################################################\n",
    "# Clean up (eg delete temp files, compress output, recompress input etc)\n",
    "#srun gzip ${input_file}\n",
    "#srun gzip ${output_file}\n",
    "echo END: `date`"
   ]
  },
  {
   "cell_type": "raw",
   "id": "7d602d05-31d0-418b-9fdf-9b595b203cfb",
   "metadata": {},
   "source": [
    "sbatch blastn_parallel.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e42545a-ccf8-47fe-87e9-d11b9eee5596",
   "metadata": {},
   "source": [
    "### check resources used:"
   ]
  },
  {
   "cell_type": "raw",
   "id": "19c044d8-fe4e-442a-ab2f-86f1c48e975d",
   "metadata": {},
   "source": [
    "sacct"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a0ae20cf-e5fd-43d7-a184-844afa98f36d",
   "metadata": {},
   "source": [
    "End of Day 1 afternoon and Day 2 afternoon sessions"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
